File: src/db/cms/Angebot.ts
import { z } from "zod";
import { client, config } from "../config";

const Angebot = z.object({
  ID: z.string(),
  Title: z.string(),
  Subtitle: z.string(),
  DateStart: z.string(),
  DateStop: z.string(),
  Link: z.string(),
  Image: z.string(),
  Anzeigen: z.boolean(),
});
type Angebot = z.infer<typeof Angebot>;

const AngebotRes = z.object({
  ID: z.string(),
  Title: z.string(),
  Subtitle: z.object({ String: z.string(), Valid: z.boolean() }),
  DateStart: z.string(),
  DateStop: z.string(),
  Link: z.string(),
  Image: z.string(),
  Anzeigen: z.object({ Bool: z.boolean(), Valid: z.boolean() }),
});
type AngebotRes = z.infer<typeof AngebotRes>;

const GetAngebotProps = z.object({ id: z.string() });
type GetAngebotProps = z.infer<typeof GetAngebotProps>;

const UpdateAngebotProps = Angebot;
type UpdateAngebotProps = z.infer<typeof UpdateAngebotProps>;

const CreateAngebotProps = z.object({
  Title: z.string(),
  Subtitle: z.string(),
  DateStart: z.string(),
  DateStop: z.string(),
  Link: z.string(),
  Image: z.string(),
  Anzeigen: z.boolean(),
});
type CreateAngebotProps = z.infer<typeof CreateAngebotProps>;

const getAngebote = async (): Promise<Angebot[] | null> => {
  const res = await client.get<AngebotRes[] | { error: string }>(
    `/CMS/Angebot`,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  const ret: Angebot[] = [];
  res.data.forEach((x) => {
    ret.push({
      Anzeigen: x.Anzeigen.Valid ? x.Anzeigen.Bool : false,
      DateStart: x.DateStart,
      DateStop: x.DateStop,
      ID: x.ID,
      Image: x.Image,
      Link: x.Link,
      Subtitle: x.Subtitle.Valid ? x.Subtitle.String : "",
      Title: x.Title,
    });
  });
  return ret;
};

const getAngebot = async (props: GetAngebotProps): Promise<Angebot | null> => {
  const res = await client.get<AngebotRes | { error: string }>(
    `/CMS/Angebot/${props.id}`,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  console.log(res.data.Subtitle);
  const x: Angebot = {
    Anzeigen: res.data.Anzeigen.Valid ? res.data.Anzeigen.Bool : false,
    DateStart: res.data.DateStart,
    DateStop: res.data.DateStop,
    ID: res.data.ID,
    Image: res.data.Image,
    Link: res.data.Link,
    Subtitle: res.data.Subtitle.Valid ? res.data.Subtitle.String : "",
    Title: res.data.Title,
  };
  return x;
};

const updateAngebot = async (props: UpdateAngebotProps): Promise<boolean> => {
  const form = new FormData();
  form.set("ID", props.ID);
  form.set("Title", props.Title);
  form.set("Subtitle", props.Subtitle);
  form.set("DateStart", props.DateStart);
  form.set("DateStop", props.DateStop);
  form.set("Link", props.Link);
  form.set("Image", props.Image);
  form.set("Anzeigen", props.Anzeigen ? "true" : "false");
  const res = await client.post<AngebotRes | { error: string }>(
    `/CMS/Angebot/${props.ID}/edit`,
    form,
    config
  );
  if ("error" in res.data) {
    return false;
  }
  return true;
};

const createAngebot = async (props: CreateAngebotProps): Promise<boolean> => {
  const form = new FormData();
  form.set("Title", props.Title);
  form.set("Subtitle", props.Subtitle);
  form.set("DateStart", props.DateStart);
  form.set("DateStop", props.DateStop);
  form.set("Link", props.Link);
  form.set("Image", props.Image);
  form.set("Anzeigen", props.Anzeigen ? "true" : "false");
  const res = await client.post<AngebotRes | { error: string }>(
    `/CMS/Angebot/new`,
    form,
    config
  );
  if ("error" in res.data) return false;

  return true;
};

const deleteAngebot = async (props: GetAngebotProps): Promise<void> => {
  const form = new FormData();
  form.set("ID", props.id);
  await client.post<{ error: string }>(
    `/CMS/Angebot/${props.id}/delete`,
    form,
    config
  );
};

export {
  getAngebot,
  getAngebote,
  updateAngebot,
  createAngebot,
  deleteAngebot,
  type Angebot,
  type GetAngebotProps,
  CreateAngebotProps,
  UpdateAngebotProps,
};

File: cmsroutes/angebot.go
package cmsroutes

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/computerextra/golang-backend/cms"
	"github.com/computerextra/golang-backend/env"
	"github.com/computerextra/golang-backend/helper"
	"github.com/google/uuid"
	"github.com/gorilla/mux"
)

func GetAngebote(w http.ResponseWriter, r *http.Request) {
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	env := env.GetEnv()
	// Get Database
	datebase, err := sql.Open("mysql", env.CMS_DATABASE_URL)
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.GetAngeboten(ctx)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func GetAngebot(w http.ResponseWriter, r *http.Request) {
	id := mux.Vars(r)["id"]
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	env := env.GetEnv()
	// Get Database
	datebase, err := sql.Open("mysql", env.CMS_DATABASE_URL)
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.GetAngebot(ctx, id)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func UpdateAngebot(w http.ResponseWriter, r *http.Request) {
	ID := r.FormValue("ID")
	Title := r.FormValue("Title")
	Subtitle := r.FormValue("Subtitle")
	DateStart := r.FormValue("DateStart") // Format: 30.09.1991
	DateStop := r.FormValue("DateStop")   // Format: 30.09.1991
	Link := r.FormValue("Link")
	Image := r.FormValue("Image")
	Anzeigen := r.FormValue("Anzeigen")
	var Start time.Time
	var Stop time.Time

	if len(DateStart) > 1 {
		split := strings.Split(DateStart, ".")
		year, err := strconv.Atoi(split[2])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		month, err := strconv.Atoi(split[1])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		day, err := strconv.Atoi(split[0])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		Start = time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	}
	if len(DateStop) > 1 {
		split := strings.Split(DateStop, ".")
		year, err := strconv.Atoi(split[2])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}
		month, err := strconv.Atoi(split[1])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}
		day, err := strconv.Atoi(split[0])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		Stop = time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	}

	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	env := env.GetEnv()
	// Get Database
	datebase, err := sql.Open("mysql", env.CMS_DATABASE_URL)
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.UpdateAngebot(ctx, cms.UpdateAngebotParams{
		Title:     Title,
		Subtitle:  sql.NullString{String: Subtitle, Valid: helper.If(len(Subtitle) > 1, true, false)},
		DateStart: Start,
		DateStop:  Stop,
		Link:      Link,
		Image:     Image,
		Anzeigen:  sql.NullBool{Bool: helper.If(Anzeigen == "true", true, false), Valid: helper.If(len(Anzeigen) > 1, true, false)},
		ID:        ID,
	})
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func CreateAngebot(w http.ResponseWriter, r *http.Request) {
	Title := r.FormValue("Title")
	Subtitle := r.FormValue("Subtitle")
	DateStart := r.FormValue("DateStart") // Format: 30.09.1991
	DateStop := r.FormValue("DateStop")   // Format: 30.09.1991
	Link := r.FormValue("Link")
	Image := r.FormValue("Image")
	Anzeigen := r.FormValue("Anzeigen")
	var Start time.Time
	var Stop time.Time

	if len(DateStart) > 1 {
		split := strings.Split(DateStart, ".")

		year, err := strconv.Atoi(split[2])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		month, err := strconv.Atoi(split[1])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		day, err := strconv.Atoi(split[0])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		Start = time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	}
	if len(DateStop) > 1 {
		split := strings.Split(DateStop, ".")
		year, err := strconv.Atoi(split[2])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}
		month, err := strconv.Atoi(split[1])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}
		day, err := strconv.Atoi(split[0])
		if err != nil {
			fehler := err.Error()
			json.NewEncoder(w).Encode(map[string]string{"error": fehler})
			return
		}

		Stop = time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	}
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	env := env.GetEnv()
	// Get Database
	datebase, err := sql.Open("mysql", env.CMS_DATABASE_URL)
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.CreateAngebot(ctx, cms.CreateAngebotParams{
		Title:     Title,
		Subtitle:  sql.NullString{String: Subtitle, Valid: helper.If(len(Subtitle) > 1, true, false)},
		DateStart: Start,
		DateStop:  Stop,
		Link:      Link,
		Image:     Image,
		Anzeigen:  sql.NullBool{Bool: helper.If(Anzeigen == "true", true, false), Valid: helper.If(len(Anzeigen) > 1, true, false)},
		ID:        uuid.New().String(),
	})
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func DeleteAngebot(w http.ResponseWriter, r *http.Request) {
	ID := r.FormValue("ID")
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	env := env.GetEnv()
	// Get Database
	datebase, err := sql.Open("mysql", env.CMS_DATABASE_URL)
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	err = queries.DeleteAngebot(ctx, ID)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(map[string]string{"error": ""})
	}
}
